Applications/spire/source/charting/chart_view.cpp

```diff
@@ -87,43 +112,70 @@ void ChartView::paintEvent(QPaintEvent* event) {
     return;
   }
   auto y_step_count = y_range / y_step;
-  auto y_pixel_step = y_origin / y_step_count;
+  auto y_pixel_step = m_y_origin / y_step_count;
   if((y_step_count - 1) * y_pixel_step < height()) {
     y_pixel_step += 1;
   }
   auto y_value = m_bottom_right.m_y;
-  auto y = y_origin;
+  auto y = m_y_origin;
   while(y_value <= m_top_left.m_y) {
     y_value += y_step;
     y -= y_pixel_step;
     painter.setPen("#3A3348");
-    painter.drawLine(0, y, x_origin, y);
+    painter.drawLine(0, y, m_x_origin, y);
     painter.setPen(Qt::white);
-    painter.drawLine(x_origin, y, x_origin + scale_width(2), y);
-    painter.drawText(x_origin + scale_width(3), y + (font_metrics.height() / 3),
+    painter.drawLine(m_x_origin, y, m_x_origin + scale_width(2), y);
+    painter.drawText(m_x_origin + scale_width(3),
+      y + (font_metrics.height() / 3),
       m_item_delegate->displayText(to_variant(m_y_axis_type, y_value),
-      QLocale()));
+        QLocale()));
   }
   auto x_step_count = x_range / x_step;
-  auto x_pixel_step = x_origin / (x_range / x_step);
+  auto x_pixel_step = m_x_origin / (x_range / x_step);
   if((x_step_count - 1) * x_pixel_step < width()) {
     x_pixel_step += 1;
   }
   auto x_text_width = font_metrics.width(get_string(m_x_axis_type,
     m_top_left.m_x));
   auto x_value = m_bottom_right.m_x;
-  auto x = x_origin;
+  auto x = m_x_origin;
   while(x_value >= m_top_left.m_x) {
     x_value -= x_step;
     x -= x_pixel_step;
     painter.setPen("#3A3348");
-    painter.drawLine(x, 0, x, y_origin);
+    painter.drawLine(x, 0, x, m_y_origin);
     painter.setPen(Qt::white);
-    painter.drawLine(x, y_origin, x, y_origin + scale_height(2));
+    painter.drawLine(x, m_y_origin, x, m_y_origin + scale_height(2));
     painter.drawText(x - x_text_width / 2,
-      y_origin + font_metrics.height() + scale_height(2),
+      m_y_origin + font_metrics.height() + scale_height(2),
       get_string(m_x_axis_type, x_value));
   }
+  if(!m_crosshair_pos.isNull()) {
+    painter.setPen(m_dashed_line_pen);
+    painter.drawLine(m_crosshair_pos.x(), 0, m_crosshair_pos.x(), m_y_origin);
+    painter.drawLine(0, m_crosshair_pos.y(), m_x_origin, m_crosshair_pos.y());
+    painter.drawImage(m_crosshair_pos.x() - (m_crosshair.width() / 2),
+      m_crosshair_pos.y() - (m_crosshair.height() / 2), m_crosshair);
+    painter.fillRect(m_crosshair_pos.x() - (scale_width(48) / 2), m_y_origin,
+      scale_width(48), scale_height(21), Qt::white);
+    painter.fillRect(m_crosshair_pos.x(), m_y_origin, scale_width(1),
+      scale_height(3), Qt::black);
+    auto crosshair_value = convert_pixels_to_chart(m_crosshair_pos);
+    m_timestamp_format = "%H:%M:%S";
```

Comment:
Don't update any member variables/state in the paint event.

Reply:
I'll have to calculate the 'step' in the axis values when the region is set/updated to implement this - and, taking that out of the paintEvent will probably speed this up as a side effect.

This is linked to the issue with the crosshair labels not lining up properly with the axis labels, so should I just close this PR and reopen it when I'm done?